File: Graphs/reachable_nodes.py
from typing import List


class ReachableNodes:
    def reachableNodes(
        self, n: int, edges: List[List[int]], restricted: List[int]
    ) -> int:
        # Pre-process graph
        graph = [[] for _ in range(n)]
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        
        # Create seen array and include restricted nodes
        seen = [False] * n
        seen[0] = True
        for node in restricted:
            seen[node] = True

        reachable = 1
        stack = [0]

        while stack:
            curr = stack.pop()
            for neighbor in graph[curr]:
                if not seen[neighbor]:
                    seen[neighbor] = True
                    reachable += 1
                    stack.append(neighbor)
        
        return reachable

# Time Complexity: O(n) where n is the number of nodes. 
# It is not O(n + e) because we are guaranteed to have n - 1 edges, so O(n + e) simplifies to O(n).
# We visit each node once and each edge once.
# Space Complexity: O(n) where n is the number of nodes.
# We use a stack to perform DFS, which in the worst case can store all nodes,
# and we use an array to keep track of visited nodes.

File: Graphs/reachableNodes.js
/*There is an undirected tree with n nodes labeled from 0 to n - 1 and n - 1 edges.

You are given a 2D integer array edges of length n - 1 where edges[i] = [ai, bi] indicates that there is an edge between nodes ai and bi in the tree. 
You are also given an integer array restricted which represents restricted nodes.

Return the maximum number of nodes you can reach from node 0 without visiting a restricted node.

Note that node 0 will not be a restricted node.*/

const reachableNodes = (n, edges, restricted) => {
  const dfs = (node) => {
    let count = 1;

    for (let neighbor of adjacencyList[node]) {
      if (!seen[neighbor]) {
        seen[neighbor] = true;
        count += dfs(neighbor);
      }
    }

    return count;
  };

  // Build adjacency list
  const adjacencyList = Array.from({ length: n }, () => []);

  for (let [nodeA, nodeB] of edges) {
    adjacencyList[nodeA].push(nodeB);
    adjacencyList[nodeB].push(nodeA);
  }

  const seen = new Array(n).fill(false);
  for (let node of restricted) {
    seen[node] = true;
  }
  seen[0] = true;

  return dfs(0);
};

// Time complexity: O(n) - We visit each node at most once.
// Space complexity: O(n) - The adjacency list and the seen array both take O(n) space.
